### ISSUE TITLE
Channel attach timeouts

### ISSUE THREAD

**Issue by mattheworiordan** [ably/ably-dotnet#104]
@marto83 a customer has reported that messages are not being published sometimes, but investigating this further it appears to be a result of a channel attach timeout.  See the logs below.

It's hard to know why one of the connections is becoming disconnected as the logs appear to show three connections being established.  I am finding out why that is being done.  If you spot anything obvious shout, otherwise I will update this ticket once I hear from the customer.

```
2017-05-29 09:39:13.6446 CustomerModule+MyLogger 01:39:13.644 #11 Protocol set to: MsgPack
2017-05-29 09:39:14.1629 CustomerModule+MyLogger 01:39:14.162 #11 Initializing message encodings. json,utf-8,cipher,base64 initialized
2017-05-29 09:39:14.2050 CustomerModule+MyLogger 01:39:14.205 #11 Authentication method: Token auth with client-side signing
2017-05-29 09:39:14.3160 CustomerModule+MyLogger 01:39:14.316 #11 xx Changing state from Initialized => Connecting. SkipAttach = False.
2017-05-29 09:39:14.3160 CustomerModule+MyLogger 01:39:14.319 #50 ConnectionManager thread created
2017-05-29 09:39:15.6500 CustomerModule+MyLogger 01:39:15.650 #50 xx Connecting: BeforeTransition
2017-05-29 09:39:15.7070 CustomerModule+MyLogger 01:39:15.707 #50 Connection notifying subscribers for state change `Connecting`
2017-05-29 09:39:15.7540 CustomerModule+MyLogger 01:39:15.754 #50 xx Connecting: Attaching state 
2017-05-29 09:39:15.7981 CustomerModule+MyLogger 01:39:15.798 #50 Creating transport
2017-05-29 09:39:15.9502 CustomerModule+MyLogger 01:39:15.950 #50 Connecting to web socket on url: wss://realtime.ably.io/?key=[REDACTED-APP-ID].JSvo2g:[REDACTED]&v=0.8&lib=dotnet-0.8.4&format=msgpack&echo=true
2017-05-29 09:39:16.0012 CustomerModule+MyLogger 01:39:16.001 #50 Connecting socket
2017-05-29 09:39:16.0963 CustomerModule+MyLogger 01:39:16.096 #50 Setting up timer 'Connecting state timer' to run action after 10 seconds. Autoreset: False
2017-05-29 09:39:16.7507 CustomerModule+MyLogger 01:39:16.750 #6 Websocket opened!
2017-05-29 09:39:16.7978 CustomerModule+MyLogger 01:39:16.797 #50 Transport state changed to: Connected.
2017-05-29 09:39:19.2455 Customer.BusinessLogic.Integrations.PushNotificationService Push notification Result: Enqueued
2017-05-29 09:39:19.4996 Customer.BusinessLogic.Integrations.PushNotificationService Push notification Result: Enqueued
2017-05-29 09:39:19.5357 CustomerModule+MyLogger 01:39:19.535 #11 #Customer:2149: Changing state to: 'Attaching'. 
2017-05-29 09:39:19.5787 CustomerModule+MyLogger 01:39:19.578 #11 Aborting timer '#Customer:2149 timer'
2017-05-29 09:39:19.6227 CustomerModule+MyLogger 01:39:19.622 #11 Setting up timer '#Customer:2149 timer' to run action after 10 seconds. Autoreset: False
2017-05-29 09:39:19.7028 CustomerModule+MyLogger 01:39:19.702 #11 Current state: Connecting. Sending message: { action=Attach }
2017-05-29 09:39:19.7288 CustomerModule+MyLogger 01:39:19.728 #11 Queuing message with action: Attach. Connection State: Connecting
2017-05-29 09:39:19.7578 CustomerModule+MyLogger 01:39:19.757 #11 #Customer:2149:Attaching queuing message
2017-05-29 09:39:20.0720 CustomerModule+MyLogger 01:39:20.072 #11 Protocol set to: MsgPack
2017-05-29 09:39:20.1051 CustomerModule+MyLogger 01:39:20.105 #11 Initializing message encodings. json,utf-8,cipher,base64 initialized
2017-05-29 09:39:20.1411 CustomerModule+MyLogger 01:39:20.141 #11 Authentication method: Token auth with client-side signing
2017-05-29 09:39:20.1701 CustomerModule+MyLogger 01:39:20.170 #11 xx Changing state from Initialized => Connecting. SkipAttach = False.
2017-05-29 09:39:20.1701 CustomerModule+MyLogger 01:39:20.170 #52 ConnectionManager thread created
2017-05-29 09:39:20.2231 CustomerModule+MyLogger 01:39:20.223 #52 xx Connecting: BeforeTransition
2017-05-29 09:39:20.2592 CustomerModule+MyLogger 01:39:20.259 #52 Connection notifying subscribers for state change `Connecting`
2017-05-29 09:39:20.2872 CustomerModule+MyLogger 01:39:20.287 #52 xx Connecting: Attaching state 
2017-05-29 09:39:20.3242 CustomerModule+MyLogger 01:39:20.324 #52 Creating transport
2017-05-29 09:39:20.3783 CustomerModule+MyLogger 01:39:20.378 #52 Connecting to web socket on url: wss://realtime.ably.io/?key=[REDACTED-APP-ID].JSvo2g:[REDACTED]&v=0.8&lib=dotnet-0.8.4&format=msgpack&echo=true
2017-05-29 09:39:20.4153 CustomerModule+MyLogger 01:39:20.415 #52 Connecting socket
2017-05-29 09:39:20.4543 CustomerModule+MyLogger 01:39:20.454 #52 Setting up timer 'Connecting state timer' to run action after 10 seconds. Autoreset: False
2017-05-29 09:39:20.7535 CustomerModule+MyLogger 01:39:20.753 #51 Websocket opened!
2017-05-29 09:39:20.7775 CustomerModule+MyLogger 01:39:20.777 #52 Transport state changed to: Connected.
2017-05-29 09:39:21.4290 Customer.BusinessLogic.Integrations.PushNotificationService Push notification Result: Enqueued
2017-05-29 09:39:21.4720 CustomerModule+MyLogger 01:39:21.472 #12 Protocol set to: MsgPack
2017-05-29 09:39:21.5010 CustomerModule+MyLogger 01:39:21.501 #12 Initializing message encodings. json,utf-8,cipher,base64 initialized
2017-05-29 09:39:21.5251 CustomerModule+MyLogger 01:39:21.525 #12 Authentication method: Token auth with client-side signing
2017-05-29 09:39:21.5661 CustomerModule+MyLogger 01:39:21.566 #12 xx Changing state from Initialized => Connecting. SkipAttach = False.
2017-05-29 09:39:21.5661 CustomerModule+MyLogger 01:39:21.570 #53 ConnectionManager thread created
2017-05-29 09:39:21.6471 Customer.BusinessLogic.Integrations.PushNotificationService Push notification Result: Enqueued
2017-05-29 09:39:21.6471 CustomerModule+MyLogger 01:39:21.649 #53 xx Connecting: BeforeTransition
2017-05-29 09:39:21.7082 CustomerModule+MyLogger 01:39:21.708 #11 #Customer:2149: Changing state to: 'Attaching'. 
2017-05-29 09:39:21.8052 CustomerModule+MyLogger 01:39:21.805 #53 Connection notifying subscribers for state change `Connecting`
2017-05-29 09:39:21.8763 CustomerModule+MyLogger 01:39:21.876 #11 Aborting timer '#Customer:2149 timer'
2017-05-29 09:39:22.0514 CustomerModule+MyLogger 01:39:22.051 #53 xx Connecting: Attaching state 
2017-05-29 09:39:22.1385 CustomerModule+MyLogger 01:39:22.138 #11 Setting up timer '#Customer:2149 timer' to run action after 10 seconds. Autoreset: False
2017-05-29 09:39:22.2556 CustomerModule+MyLogger 01:39:22.255 #53 Creating transport
2017-05-29 09:39:22.2816 CustomerModule+MyLogger 01:39:22.281 #11 Current state: Connecting. Sending message: { action=Attach }
2017-05-29 09:39:22.3106 CustomerModule+MyLogger 01:39:22.310 #53 Connecting to web socket on url: wss://realtime.ably.io/?key=[REDACTED-APP-ID].JSvo2g:[REDACTED]&v=0.8&lib=dotnet-0.8.4&format=msgpack&echo=true
2017-05-29 09:39:22.3376 CustomerModule+MyLogger 01:39:22.337 #11 Queuing message with action: Attach. Connection State: Connecting
2017-05-29 09:39:22.3927 CustomerModule+MyLogger 01:39:22.392 #53 Connecting socket
2017-05-29 09:39:22.4227 CustomerModule+MyLogger 01:39:22.422 #11 #Customer:2149:Attaching queuing message
2017-05-29 09:39:22.4577 CustomerModule+MyLogger 01:39:22.457 #53 Setting up timer 'Connecting state timer' to run action after 10 seconds. Autoreset: False
2017-05-29 09:39:22.7709 CustomerModule+MyLogger 01:39:22.770 #4 Websocket opened!
2017-05-29 09:39:22.8300 CustomerModule+MyLogger 01:39:22.830 #6 Websocket data message received. Raw: { "action" : 4, "connectionId" : "_o8qp9RQJf", "connectionKey" : "REDACTED", "connectionSerial" : -1, "connectionDetails" : { "clientId" : "*", "connectionKey" : "REDACTED", "maxMessageSize" : 65536, "maxInboundRate" : 50, "maxFrameSize" : 262144, "serverId" : "frontend.e8ff.1.ap-southeast-2-A.i-06a4db4f1c287111a", "connectionStateTtl" : 120000, "maxIdleInterval" : 15000 } }
2017-05-29 09:39:22.8530 CustomerModule+MyLogger 01:39:22.853 #53 Transport state changed to: Connected.
2017-05-29 09:39:22.9570 CustomerModule+MyLogger 01:39:22.957 #52 Aborting timer 'Connecting state timer'
2017-05-29 09:39:22.9971 CustomerModule+MyLogger 01:39:22.997 #52 xx Changing state from Connecting => Connected. SkipAttach = False.
2017-05-29 09:39:23.0121 Customer.BusinessLogic.Integrations.PushNotificationService Push notification Result: Enqueued
2017-05-29 09:39:23.0121 CustomerModule+MyLogger 01:39:23.024 #52 Aborting timer 'Connecting state timer'
2017-05-29 09:39:23.0741 CustomerModule+MyLogger 01:39:23.074 #52 xx Connected: BeforeTransition
2017-05-29 09:39:23.1011 CustomerModule+MyLogger 01:39:23.101 #52 Connection notifying subscribers for state change `Connected`
2017-05-29 09:39:23.1282 CustomerModule+MyLogger 01:39:23.128 #52 xx Connected: Attaching state 
2017-05-29 09:39:23.1932 CustomerModule+MyLogger 01:39:23.193 #52 Processing Connected:OnAttached. Resumed: 
2017-05-29 09:39:23.2182 CustomerModule+MyLogger 01:39:23.218 #52 Sending pending message: Count: 1
2017-05-29 09:39:23.2182 Customer.BusinessLogic.Integrations.PushNotificationService Push notification Result: Enqueued
2017-05-29 09:39:23.2452 CustomerModule+MyLogger 01:39:23.245 #52 Sending message (Attach) to transport
2017-05-29 09:39:23.2723 CustomerModule+MyLogger 01:39:23.272 #12 #Customer:2149: Changing state to: 'Attaching'. 
2017-05-29 09:39:23.3053 CustomerModule+MyLogger 01:39:23.305 #52 Transport state (Open): Sending message. Action: Attach - Binary message with length: 39
2017-05-29 09:39:23.3203 CustomerModule+MyLogger 01:39:23.320 #12 Aborting timer '#Customer:2149 timer'
2017-05-29 09:39:23.3924 CustomerModule+MyLogger 01:39:23.392 #12 Setting up timer '#Customer:2149 timer' to run action after 10 seconds. Autoreset: False
2017-05-29 09:39:23.4324 CustomerModule+MyLogger 01:39:23.432 #12 Current state: Connecting. Sending message: { action=Attach }
2017-05-29 09:39:23.4564 CustomerModule+MyLogger 01:39:23.456 #12 Queuing message with action: Attach. Connection State: Connecting
2017-05-29 09:39:23.4904 CustomerModule+MyLogger 01:39:23.490 #12 #Customer:2149:Attaching queuing message
2017-05-29 09:39:24.3930 CustomerModule+MyLogger 01:39:24.393 #51 Websocket data message received. Raw: { "action" : 4, "connectionId" : "KKNCBaTi_V", "connectionKey" : "REDACTED", "connectionSerial" : -1, "connectionDetails" : { "clientId" : "*", "connectionKey" : "REDACTED", "maxMessageSize" : 65536, "maxInboundRate" : 50, "maxFrameSize" : 262144, "serverId" : "frontend.6e97.2.ap-southeast-2-A.i-0f85e0b09173e2e7f", "connectionStateTtl" : 120000, "maxIdleInterval" : 15000 } }
2017-05-29 09:39:24.4231 CustomerModule+MyLogger 01:39:24.423 #53 Aborting timer 'Connecting state timer'
2017-05-29 09:39:24.4471 CustomerModule+MyLogger 01:39:24.447 #53 xx Changing state from Connecting => Connected. SkipAttach = False.
2017-05-29 09:39:24.4931 CustomerModule+MyLogger 01:39:24.493 #53 Aborting timer 'Connecting state timer'
2017-05-29 09:39:24.5291 CustomerModule+MyLogger 01:39:24.529 #53 xx Connected: BeforeTransition
2017-05-29 09:39:24.5692 CustomerModule+MyLogger 01:39:24.569 #53 Connection notifying subscribers for state change `Connected`
2017-05-29 09:39:24.6122 CustomerModule+MyLogger 01:39:24.612 #53 xx Connected: Attaching state 
2017-05-29 09:39:24.6392 CustomerModule+MyLogger 01:39:24.639 #53 Processing Connected:OnAttached. Resumed: 
2017-05-29 09:39:24.6732 CustomerModule+MyLogger 01:39:24.673 #53 Sending pending message: Count: 1
2017-05-29 09:39:24.7033 CustomerModule+MyLogger 01:39:24.703 #53 Sending message (Attach) to transport
2017-05-29 09:39:24.7443 CustomerModule+MyLogger 01:39:24.744 #53 Transport state (Open): Sending message. Action: Attach - Binary message with length: 39
2017-05-29 09:39:25.7880 CustomerModule+MyLogger 01:39:25.788 #51 Websocket data message received. Raw: { "action" : 11, "flags" : 0, "channel" : "Customer:2149", "channelSerial" : "610fJGkiAA7YU585463072:-1" }
2017-05-29 09:39:25.7880 CustomerModule+MyLogger 01:39:25.789 #49 Websocket data message received. Raw: { "action" : 11, "flags" : 0, "channel" : "Customer:2149", "channelSerial" : "610fJGkiAA7YU585463072:-1" }
2017-05-29 09:39:25.8210 CustomerModule+MyLogger 01:39:25.821 #52 #Customer:2149: Changing state to: 'Attached'. 
2017-05-29 09:39:25.8541 CustomerModule+MyLogger 01:39:25.854 #53 #Customer:2149: Changing state to: 'Attached'. 
2017-05-29 09:39:25.8831 CustomerModule+MyLogger 01:39:25.883 #52 Aborting timer '#Customer:2149 timer'
2017-05-29 09:39:25.9081 CustomerModule+MyLogger 01:39:25.908 #53 Aborting timer '#Customer:2149 timer'
2017-05-29 09:39:25.9311 CustomerModule+MyLogger 01:39:25.931 #52 Current state: Connected. Sending message: { action=Message, mesasages=[ { name="CreatedCustomerItem", data={"createdUserId":"9","createdUserFirstname":"First","createdUserLastname":"Last","id":"7407","type":"TEXT","url":null,"text":"2","isEdited":false,"thumbnailUrl":null,"imageUrl":null,"downloadUrl":null,"videoUrl":null,"latitude":null,"longitude":null,"createDateTime":"2017-05-29T01:39:20.6504937Z","lastUpdateDateTime":"2017-05-29T01:39:20.6504937Z","microthreds":[],"profilePictureUrl":"http://127.0.0.1:10000/devstoreaccount1/images/1821d2c5-2ce5-4118-88b3-43a32295742d.png?sv=2015-12-11&sr=b&sig=Hse5q45%2FEtzPd1elFLVqIxbD05j8otEYtSKihjXwOtA%3D&se=9999-12-31T15%3A59%3A59Z&sp=r","additionalContent":null,"firstUnreadCustomerItem":false,"channelItemId":null,"channelId":null}} ] }
2017-05-29 09:39:25.9501 CustomerModule+MyLogger 01:39:25.950 #53 Current state: Connected. Sending message: { action=Message, mesasages=[ { name="CreatedCustomerItem", data={"createdUserId":"9","createdUserFirstname":"First","createdUserLastname":"Last","id":"7408","type":"TEXT","url":null,"text":"3","isEdited":false,"thumbnailUrl":null,"imageUrl":null,"downloadUrl":null,"videoUrl":null,"latitude":null,"longitude":null,"createDateTime":"2017-05-29T01:39:22.0924958Z","lastUpdateDateTime":"2017-05-29T01:39:22.0924958Z","microthreds":[],"profilePictureUrl":"http://127.0.0.1:10000/devstoreaccount1/images/1821d2c5-2ce5-4118-88b3-43a32295742d.png?sv=2015-12-11&sr=b&sig=Hse5q45%2FEtzPd1elFLVqIxbD05j8otEYtSKihjXwOtA%3D&se=9999-12-31T15%3A59%3A59Z&sp=r","additionalContent":null,"firstUnreadCustomerItem":false,"channelItemId":null,"channelId":null}} ] }
2017-05-29 09:39:25.9842 CustomerModule+MyLogger 01:39:25.984 #52 Message (Message) with serial (0) was queued to get Ack
2017-05-29 09:39:26.0472 CustomerModule+MyLogger 01:39:26.047 #53 Message (Message) with serial (0) was queued to get Ack
2017-05-29 09:39:26.0832 CustomerModule+MyLogger 01:39:26.083 #52 Sending message (Message) to transport
2017-05-29 09:39:26.1182 CustomerModule+MyLogger 01:39:26.118 #53 Sending message (Message) to transport
2017-05-29 09:39:26.1583 CustomerModule+MyLogger 01:39:26.158 #52 Transport state (Open): Sending message. Action: Message - Binary message with length: 752
2017-05-29 09:39:26.1873 CustomerModule+MyLogger 01:39:26.187 #11 Timer 'Connecting state timer' interval 10 seconds elapsed and calling action.
2017-05-29 09:39:26.1803 CustomerModule+MyLogger 01:39:26.180 #53 Transport state (Open): Sending message. Action: Message - Binary message with length: 752
2017-05-29 09:39:26.2283 CustomerModule+MyLogger 01:39:26.228 #50 Handling Connecting failure.
2017-05-29 09:39:26.2633 CustomerModule+MyLogger 01:39:26.263 #49 Websocket data message received. Raw: { "action" : 15, "id" : "_o8qp9RQJf:0", "connectionId" : "_o8qp9RQJf", "connectionSerial" : 0, "channel" : "Customer:2149", "channelSerial" : "610fJGkiAA7YU585463072:0", "timestamp" : 1496021966555, "messages" : [ { "name" : "CreatedCustomerItem", "data" : "{\"createdUserId\":\"9\",\"createdUserFirstname\":\"First\",\"createdUserLastname\":\"Last\",\"id\":\"7407\",\"type\":\"TEXT\",\"url\":null,\"text\":\"2\",\"isEdited\":false,\"thumbnailUrl\":null,\"imageUrl\":null,\"downloadUrl\":null,\"videoUrl\":null,\"latitude\":null,\"longitude\":null,\"createDateTime\":\"2017-05-29T01:39:20.6504937Z\",\"lastUpdateDateTime\":\"2017-05-29T01:39:20.6504937Z\",\"microthreds\":[],\"profilePictureUrl\":\"http:\/\/127.0.0.1:10000\/REDACTED",\"additionalContent\":null,\"firstUnreadCustomerItem\":false,\"channelItemId\":null,\"channelId\":null}" } ] }
2017-05-29 09:39:26.2814 CustomerModule+MyLogger 01:39:26.281 #50 xx Changing state from Connecting => Disconnected. SkipAttach = False.
2017-05-29 09:39:26.3054 CustomerModule+MyLogger 01:39:26.305 #49 Websocket data message received. Raw: { "action" : 1, "count" : 1, "msgSerial" : 0 }
2017-05-29 09:39:26.3284 CustomerModule+MyLogger 01:39:26.328 #50 Aborting timer 'Connecting state timer'
2017-05-29 09:39:26.3524 CustomerModule+MyLogger 01:39:26.352 #49 Websocket data message received. Raw: { "action" : 15, "id" : "KKNCBaTi_V:0", "connectionId" : "KKNCBaTi_V", "connectionSerial" : 1, "channel" : "Customer:2149", "channelSerial" : "610fJGkiAA7YU585463072:1", "timestamp" : 1496021966599, "messages" : [ { "name" : "CreatedCustomerItem", "data" : "{\"createdUserId\":\"9\",\"createdUserFirstname\":\"First\",\"createdUserLastname\":\"Last\",\"id\":\"7408\",\"type\":\"TEXT\",\"url\":null,\"text\":\"3\",\"isEdited\":false,\"thumbnailUrl\":null,\"imageUrl\":null,\"downloadUrl\":null,\"videoUrl\":null,\"latitude\":null,\"longitude\":null,\"createDateTime\":\"2017-05-29T01:39:22.0924958Z\",\"lastUpdateDateTime\":\"2017-05-29T01:39:22.0924958Z\",\"microthreds\":[],\"profilePictureUrl\":\"http:\/\/127.0.0.1:10000\/REDACTED",\"additionalContent\":null,\"firstUnreadCustomerItem\":false,\"channelItemId\":null,\"channelId\":null}" } ] }
2017-05-29 09:39:26.3814 CustomerModule+MyLogger 01:39:26.381 #50 xx Disconnected: BeforeTransition
2017-05-29 09:39:26.2633 CustomerModule+MyLogger 01:39:26.265 #51 Websocket data message received. Raw: { "action" : 15, "id" : "_o8qp9RQJf:0", "connectionId" : "_o8qp9RQJf", "connectionSerial" : 0, "channel" : "Customer:2149", "channelSerial" : "610fJGkiAA7YU585463072:0", "timestamp" : 1496021966555, "messages" : [ { "name" : "CreatedCustomerItem", "data" : "{\"createdUserId\":\"9\",\"createdUserFirstname\":\"First\",\"createdUserLastname\":\"Last\",\"id\":\"7407\",\"type\":\"TEXT\",\"url\":null,\"text\":\"2\",\"isEdited\":false,\"thumbnailUrl\":null,\"imageUrl\":null,\"downloadUrl\":null,\"videoUrl\":null,\"latitude\":null,\"longitude\":null,\"createDateTime\":\"2017-05-29T01:39:20.6504937Z\",\"lastUpdateDateTime\":\"2017-05-29T01:39:20.6504937Z\",\"microthreds\":[],\"profilePictureUrl\":\"http:\/\/127.0.0.1:10000\/REDACTED",\"additionalContent\":null,\"firstUnreadCustomerItem\":false,\"channelItemId\":null,\"channelId\":null}" } ] }
2017-05-29 09:39:26.4485 CustomerModule+MyLogger 01:39:26.448 #50 Connection notifying subscribers for state change `Disconnected`
2017-05-29 09:39:26.4895 CustomerModule+MyLogger 01:39:26.489 #51 Websocket data message received. Raw: { "action" : 15, "id" : "KKNCBaTi_V:0", "connectionId" : "KKNCBaTi_V", "connectionSerial" : 1, "channel" : "Customer:2149", "channelSerial" : "610fJGkiAA7YU585463072:1", "timestamp" : 1496021966599, "messages" : [ { "name" : "CreatedCustomerItem", "data" : "{\"createdUserId\":\"9\",\"createdUserFirstname\":\"First\",\"createdUserLastname\":\"Last\",\"id\":\"7408\",\"type\":\"TEXT\",\"url\":null,\"text\":\"3\",\"isEdited\":false,\"thumbnailUrl\":null,\"imageUrl\":null,\"downloadUrl\":null,\"videoUrl\":null,\"latitude\":null,\"longitude\":null,\"createDateTime\":\"2017-05-29T01:39:22.0924958Z\",\"lastUpdateDateTime\":\"2017-05-29T01:39:22.0924958Z\",\"microthreds\":[],\"profilePictureUrl\":\"http:\/\/127.0.0.1:10000\/REDACTED",\"additionalContent\":null,\"firstUnreadCustomerItem\":false,\"channelItemId\":null,\"channelId\":null}" } ] }
2017-05-29 09:39:26.5145 CustomerModule+MyLogger 01:39:26.514 #50 xx Disconnected: Attaching state 
2017-05-29 09:39:26.5405 CustomerModule+MyLogger 01:39:26.540 #51 Websocket data message received. Raw: { "action" : 1, "count" : 1, "msgSerial" : 0 }
2017-05-29 09:39:26.5756 CustomerModule+MyLogger 01:39:26.575 #50 Destroying transport
2017-05-29 09:39:26.3554 Customer.Api.Controllers.CustomerItemController Success Ably Create(Customer:2149) - 7407
2017-05-29 09:39:26.6216 CustomerModule+MyLogger 01:39:26.621 #50 Closing socket. Current socket is not null
2017-05-29 09:39:26.5996 Customer.Api.Controllers.CustomerItemController Success Ably Create(Customer:2149) - 7408
2017-05-29 09:39:26.6906 CustomerModule+MyLogger 01:39:26.690 #50 Setting up timer 'Disconnected state timer' to run action after 15 seconds. Autoreset: False
2017-05-29 09:39:29.7147 CustomerModule+MyLogger 01:39:29.714 #42 Timer '#Customer:2149 timer' interval 10 seconds elapsed and calling action.
2017-05-29 09:39:29.7378 CustomerModule+MyLogger 01:39:29.737 #50 #Customer:2149: Changing state to: 'Failed'. Error: Reason: Channel didn't attach within the default timeout; Code: 50000
2017-05-29 09:39:29.7618 Customer.Api.Controllers.CustomerItemController Failed Ably Create(Customer:2149) - 7406 - Reason: Channel didn't attach within the default timeout; Code: 50000
2017-05-29 09:39:37.4511 CustomerModule+MyLogger 01:39:37.451 #51 Websocket data message received. Raw: { "action" : 0 }
2017-05-29 09:39:39.4205 CustomerModule+MyLogger 01:39:39.420 #54 Websocket data message received. Raw: { "action" : 0 }
2017-05-29 09:39:41.7631 CustomerModule+MyLogger 01:39:41.763 #41 Timer 'Disconnected state timer' interval 15 seconds elapsed and calling action.
2017-05-29 09:39:41.7961 CustomerModule+MyLogger 01:39:41.796 #50 xx Changing state from Disconnected => Connecting. SkipAttach = False.
2017-05-29 09:39:41.8192 CustomerModule+MyLogger 01:39:41.819 #50 Aborting timer 'Disconnected state timer'
2017-05-29 09:39:41.8432 CustomerModule+MyLogger 01:39:41.843 #50 xx Connecting: BeforeTransition
2017-05-29 09:39:41.8772 CustomerModule+MyLogger 01:39:41.877 #50 Connection notifying subscribers for state change `Connecting`
2017-05-29 09:39:41.9272 CustomerModule+MyLogger 01:39:41.927 #50 xx Connecting: Attaching state 
2017-05-29 09:39:41.9603 CustomerModule+MyLogger 01:39:41.960 #50 Creating transport
2017-05-29 09:39:41.9913 CustomerModule+MyLogger 01:39:41.991 #50 Connecting to web socket on url: wss://realtime.ably.io/?key=[REDACTED-APP-ID].JSvo2g:[REDACTED]&v=0.8&lib=dotnet-0.8.4&format=msgpack&echo=true
2017-05-29 09:39:42.0193 CustomerModule+MyLogger 01:39:42.019 #50 Connecting socket
2017-05-29 09:39:42.0443 CustomerModule+MyLogger 01:39:42.044 #50 Setting up timer 'Connecting state timer' to run action after 10 seconds. Autoreset: False
2017-05-29 09:39:42.3806 CustomerModule+MyLogger 01:39:42.380 #51 Websocket opened!
2017-05-29 09:39:42.4136 CustomerModule+MyLogger 01:39:42.413 #50 Transport state changed to: Connected.
2017-05-29 09:39:42.4136 CustomerModule+MyLogger 01:39:42.413 #51 Websocket data message received. Raw: { "action" : 4, "connectionId" : "ftVPISJjTx", "connectionKey" : "REDACTED", "connectionSerial" : -1, "connectionDetails" : { "clientId" : "*", "connectionKey" : "REDACTED", "maxMessageSize" : 65536, "maxInboundRate" : 50, "maxFrameSize" : 262144, "serverId" : "frontend.e8ff.1.ap-southeast-2-A.i-06a4db4f1c287111a", "connectionStateTtl" : 120000, "maxIdleInterval" : 15000 } }
2017-05-29 09:39:42.4676 CustomerModule+MyLogger 01:39:42.467 #50 Aborting timer 'Connecting state timer'
2017-05-29 09:39:42.5146 CustomerModule+MyLogger 01:39:42.514 #50 xx Changing state from Connecting => Connected. SkipAttach = False.
2017-05-29 09:39:42.5827 CustomerModule+MyLogger 01:39:42.582 #50 Aborting timer 'Connecting state timer'
2017-05-29 09:39:42.6157 CustomerModule+MyLogger 01:39:42.615 #50 xx Connected: BeforeTransition
2017-05-29 09:39:42.6397 CustomerModule+MyLogger 01:39:42.639 #50 Connection notifying subscribers for state change `Connected`
2017-05-29 09:39:42.6668 CustomerModule+MyLogger 01:39:42.666 #50 xx Connected: Attaching state 
2017-05-29 09:39:42.7048 CustomerModule+MyLogger 01:39:42.704 #50 Processing Connected:OnAttached. Resumed: 
2017-05-29 09:39:42.7558 CustomerModule+MyLogger 01:39:42.755 #50 Sending pending message: Count: 1
2017-05-29 09:39:42.7878 CustomerModule+MyLogger 01:39:42.787 #50 Sending message (Attach) to transport
2017-05-29 09:39:42.8179 CustomerModule+MyLogger 01:39:42.817 #50 Transport state (Open): Sending message. Action: Attach - Binary message with length: 39
2017-05-29 09:39:42.9249 CustomerModule+MyLogger 01:39:42.924 #51 Websocket data message received. Raw: { "action" : 11, "flags" : 0, "channel" : "Customer:2149", "channelSerial" : "610fJGkiAA7YU585463072:1" }
2017-05-29 09:39:42.9510 CustomerModule+MyLogger 01:39:42.951 #50 #Customer:2149: Changing state to: 'Attached'. 
2017-05-29 09:39:42.9740 CustomerModule+MyLogger 01:39:42.974 #50 Aborting timer '#Customer:2149 timer'
2017-05-29 09:39:52.4496 CustomerModule+MyLogger 01:39:52.449 #51 Websocket data message received. Raw: { "action" : 0 }
2017-05-29 09:39:54.3950 CustomerModule+MyLogger 01:39:54.395 #54 Websocket data message received. Raw: { "action" : 0 }
2017-05-29 09:39:57.3550 CustomerModule+MyLogger 01:39:57.355 #54 Websocket data message received. Raw: { "action" : 0 }
```

**Comment by marto83** [comment]
@mattheworiordan the logs show that the realtime library is initialised 3 times. There must be a race condition with the 'Connection state timer' when multiple instances of the library are running at the same time. 
I'm going to try and recreate it in a test and get it fixed. In the meanwhile can you ask the customer if they can use one instance of the ably realtime library.

**Comment by ChRonX13** [comment]
Hi, customer in question :) to clarify your statement, is the intention to use a 'singleton' instance of IRealtimeClient as opposed to resolving a new instance for each web request? It was also mentioned for us to try using the IRestClient instead as this is for server side publishing, is there a preferred practice?

**Comment by marto83** [comment]
Hi @ChRonX13. Regarding the Realtime client. The best practice is to use it as a singleton or application because it vmanages a websocket connection to the server. The RestClient on the other hand you can use either way - as a singleton or creating a new instance per request. All you save with the singleton is the setup for the internal HttpClient and a bit to manage the channels collection. 
Definitely use the RestClient in serverside apps.

**Comment by ChRonX13** [comment]
@marto83 Awesome...thanks a lot for that insight...we've updated our code accordingly to the Rest client and so far the team hasn't seen any issues.

**Comment by marto83** [comment]
Very welcome. Let me know if you have more issues.

**Comment by mattheworiordan** [comment]
@marto83 should we be closing this?  Whilst the issue is arguably one of how the lib was used, it's not technically incorrect to have multiple libs instanced at the same time. If the lib fails, that indicates a sharing of state between instances that is probably a bug no?

**Comment by marto83** [comment]
Good point. Let's keep it open until I find out what's causing it.

**Comment by marto83** [comment]
I updated the attaching timer and added a test to validate the issue. All good now. 
It's fixed in 0.8.5-beta12.
